pkg/controller/volume/attach_detach_controller.go

```diff
@@ -106,30 +240,367 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
 		go adc.internalNodeInformer.Run(stopCh)
 	}
 
+	if adc.selfCreatedPVCInformer {
+		go adc.internalPVCInformer.Run(stopCh)
+	}
+
+	if adc.selfCreatedPVInformer {
+		go adc.internalPVInformer.Run(stopCh)
+	}
+
+	go adc.reconciler.Run(stopCh)
+
 	<-stopCh
 	glog.Infof("Shutting down Attach Detach Controller")
 }
 
 func (adc *attachDetachController) podAdd(obj interface{}) {
-	// No op for now
+	pod, ok := obj.(*api.Pod)
+	if pod == nil || !ok {
+		return
+	}
+
+	addVolumes := pod.Status.Phase == api.PodPending ||
+		pod.Status.Phase == api.PodRunning
+	adc.processPodVolumes(pod, addVolumes /* addVolumes */)
 }
 
 func (adc *attachDetachController) podUpdate(oldObj, newObj interface{}) {
-	// No op for now
+	// The flow for update is the same as add.
+	adc.podAdd(newObj)
 }
 
 func (adc *attachDetachController) podDelete(obj interface{}) {
-	// No op for now
+	pod, ok := obj.(*api.Pod)
+	if pod == nil || !ok {
+		return
+	}
+
+	adc.processPodVolumes(pod, false /* addVolumes */)
 }
 
 func (adc *attachDetachController) nodeAdd(obj interface{}) {
-	// No op for now
+	node, ok := obj.(*api.Node)
+	if node == nil || !ok {
+		return
+	}
+
+	nodeName := node.Name
+	if _, exists := node.Annotations[ControllerManagedAnnotation]; exists {
+		// Node specifies annotation indicating it should be managed by attach
+		// detach controller. Add it to desired state of world.
+		adc.desiredStateOfWorld.AddNode(nodeName)
+	}
+
+	// Process "safe to detach" volumes annotations. Mark volumes in "actual
+	// state of world" as safe to detach. Delete annotations for the rest.
+	adc.processSafeToDetachAnnotations(nodeName, node.Annotations)
 }
 
 func (adc *attachDetachController) nodeUpdate(oldObj, newObj interface{}) {
-	// No op for now
+	// The flow for update is the same as add.
+	adc.nodeAdd(newObj)
 }
 
 func (adc *attachDetachController) nodeDelete(obj interface{}) {
-	// No op for now
+	node, ok := obj.(*api.Node)
+	if node == nil || !ok {
+		return
+	}
+
+	nodeName := node.Name
+	if err := adc.desiredStateOfWorld.DeleteNode(nodeName); err != nil {
+		glog.V(10).Infof("%v", err)
+	}
+
+	// Process "safe to detach" volumes annotations. Mark volumes in "actual
+	// state of world" as safe to detach. Delete annotations for the rest.
+	adc.processSafeToDetachAnnotations(nodeName, node.Annotations)
```

Comment:
So if a node is deleted should we not detach all volumes ?.. ie mark all its volumes SafeToDetach ?